pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -223,56 +300,54 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
 	return p, nil
 }
 
-// AddAllowed adds reStr, a regexp, to the DNS lookups the proxy allows.
-func (p *DNSProxy) AddAllowed(reStr, endpointID string) {
-	log.WithField(logfields.DNSName, reStr).Debug("Adding allowed DNS FQDN pattern")
-	p.UpdateAllowed([]string{reStr}, nil, endpointID)
-}
+// UpdateAllowed sets newRules for endpointID and destPort. It compiles the DNS
+// rules into regexes that are then used in CheckAllowed.
+func (p *DNSProxy) UpdateAllowed(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
+	p.Lock()
+	defer p.Unlock()
 
-// RemoveAllowed removes reStr from the DNS lookups the proxy allows. It must
-// match the form in AddAllowed exactly (i.e. this isn't removing by regex, but
-// by direct equivalence).
-func (p *DNSProxy) RemoveAllowed(reStr, endpointID string) {
-	log.WithField(logfields.DNSName, reStr).Debug("Removing allowed DNS FQDN pattern")
-	p.UpdateAllowed(nil, []string{reStr}, endpointID)
+	return p.allowed.setPortRulesForID(endpointID, destPort, newRules)
 }
 
-// UpdateAllowed adds and removes reStr while holding the lock. This is a bit
-// of a hack to ensure atomic updates of rules until we replace the tracking
-// with something better.
-func (p *DNSProxy) UpdateAllowed(reStrToAdd, reStrToRemove []string, endpointID string) {
-	for i := range reStrToAdd {
-		reStrToAdd[i] = prepareNameMatch(reStrToAdd[i])
-	}
-	for i := range reStrToRemove {
-		reStrToRemove[i] = prepareNameMatch(reStrToRemove[i])
-	}
-
+// CheckAllowed checks endpointID, destPort, destID, and name against the rules
+// added to the proxy, and only returns true if this all match something that
+// was added (via SetAllowed) previously.
+func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID identity.NumericIdentity, name string) (allowed bool, err error) {
+	name = strings.ToLower(dns.Fqdn(name))
 	p.Lock()
 	defer p.Unlock()
-	for _, reStr := range reStrToRemove {
-		p.allowed.Remove(reStr, endpointID)
+
+	epAllow, exists := p.allowed.getPortRulesForID(endpointID, destPort)
+	if !exists {
+		return false, nil
 	}
-	for _, reStr := range reStrToAdd {
-		p.allowed.Add(reStr, endpointID)
+
+	// FIXME: We see cachedSelectors with nil internals, possibly from visibility
```

Comment:
@joestringer I'm a bit unsure about this. This is my best guess, but it feels like a hack. Is there a way to check here that it's a visibility policy? (instead of allowing blindly on a panic)

Reply:
Here's your problem:
https://github.com/cilium/cilium/blob/904b274e2c76edf232758126cce4b369bf356a3a/pkg/policy/visibility.go#L148

This looks like it should be `api.WildcardEndpointSelector` instead.

Reply:
Hmm, it wants a type that implements CachedSelector (this is missing GetSelections it seems). I'll try to figure out how to build the right objects. I'm not 100% clear on the path from one to the other.

Reply:
Ah, right you are. Maybe we should consider creating a singleton `CachedWildcardEndpointSelector` which caches the selections of the `api.WildcardEndpointSelector` for use here.

Reply:
Yup, that's exactly what I tried. I wasn't sure what to pass into the constructor, though. It want a "user". I'll look through the way we make others and figure it out. I agree that it's the righter thing to do (over the panic/recover, certainly).

Reply:
@joestringer Ok, I couldn't do it quite right but it self-consistent. The selector realises that it's a wildcard, and so avoids dereferencing itself (well, parts). It's a tad brittle because it relies on the current internal logic but I doubt that part will change (short-circuiting on wildcard).
I couldn't make this a global side-by-side with `api.WildcardEndpointSelector` because it relies on the policy-internal `wildcardSelectorKey`. I can move those about more if you think it's worth it. I can also make this part a separate PR

Reply:
The `user` part is the mechanism we would use when a new identity shows up in the cluster which matches the 'policy' so that we can push an incremental update down to the datapath.

The nice thing with the visibility policy is that it is supposed to match all traffic, so I think even if the exact plumbing is subtly different, this is likely to correctly match traffic. Looking at what you've done, it looks sufficient for now. We can take a note to explore more subtle cases like identity update to ensure that works as expected, as a follow-up to this PR.